File: TrainsProblem/GraphComputation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace TrainsProblem
{
    public static class GraphComputation
    {
        public static List<List<Node>> FindAllPaths(Node from, Node to)
        {
            var nodes = new Queue<Tuple<Node, List<Node>>>();
            nodes.Enqueue(new Tuple<Node, List<Node>>(from, new List<Node>()));
            var paths = new List<List<Node>>();

            while (nodes.Any())
            {
                var current = nodes.Dequeue();
                Node currentNode = current.Item1;

                if (current.Item2.Contains(currentNode))
                {
                    continue;
                }

                current.Item2.Add(currentNode);

                if (currentNode == to)
                {
                    paths.Add(current.Item2);
                    continue;
                }

                foreach (var edge in current.Item1.Edges)
                {
                    nodes.Enqueue(new Tuple<Node, List<Node>>(edge.NodeDestination, new List<Node>(current.Item2)));
                }
            }

            return paths;
        }

        public static GraphResult GetAllPathInfo(Node from, Node to)
        {
            var result = new GraphResult()
            {
                Start = from,
                End = to
            };

            var allPaths = FindAllPaths(from, to).ToList();

            var pathInfo = new RouteInfo();
            foreach (var path in allPaths)
            {
                pathInfo = GetPathInfo(path);
                result.PossibleRoutes.Add(pathInfo);
            }

            return result;
        }

        public static RouteInfo GetPathInfo(List<Node> paths)
        {
            //get the number of stops on each path
            var route = new StringBuilder();
            int totalDistance = 0;
            int counter = 0;
            var initialNode = new Node();

            foreach (Node stopOver in paths)
            {
                // set route map
                route.AppendFormat("{0} ", stopOver.Name);

                //get distance between each other
                if (counter < paths.Count)
                {
                    if (counter > 0) //check if first item
                    {
                        //compare
                        Edge endPoint = initialNode.Edges.FirstOrDefault(e => e.NodeDestination.Name == stopOver.Name);

                        if (endPoint == null)
                        {
                            return null;
                        }

                        //get distance
                        totalDistance += endPoint.Distance;
                    }

                    //replace initial node; will be used to compare with the next node
                    initialNode = stopOver;
                    counter++;
                }
            }

            var result = new RouteInfo
            {
                Path = route.ToString(),
                NumberOfStops = paths.Count,
                TotalDistance = totalDistance
            };

            return result;
        }
    }
}

File: TrainsProblem.VB/GraphComputation.vb

Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace TrainsProblem
    Public NotInheritable Class GraphComputation
        Private Sub New()
        End Sub
        Public Shared Function FindAllPaths(from As Node, [to] As Node) As List(Of List(Of Node))
           
            Dim nodes = New Queue(Of Tuple(Of Node, List(Of Node)))()
            nodes.Enqueue(New Tuple(Of Node, List(Of Node))(from, New List(Of Node)()))
            Dim paths = New List(Of List(Of Node))()

            While nodes.Any()
                Dim current = nodes.Dequeue()
                Dim currentNode As Node = current.Item1

                If current.Item2.Contains(currentNode) Then
                    Continue While
                End If

                current.Item2.Add(currentNode)

                If currentNode.Equals([to]) Then
                    paths.Add(current.Item2)
                    Continue While
                End If

                For Each edge As Edge In current.Item1.Edges
                    nodes.Enqueue(New Tuple(Of Node, List(Of Node))(edge.NodeDestination, New List(Of Node)(current.Item2)))
                Next
            End While

            Return paths
        End Function

        Public Shared Function GetAllPathInfo(from As Node, [to] As Node) As GraphResult
            Dim result = New GraphResult() With { _
                .Start = from, _
                .[End] = [to] _
            }

            Dim allPaths = FindAllPaths(from, [to]).ToList()

            Dim pathInfo = New RouteInfo()
            For Each path As List(Of Node) In allPaths
                pathInfo = GetPathInfo(path)
                result.PossibleRoutes.Add(pathInfo)
            Next

            Return result
        End Function

        Public Shared Function GetPathInfo(paths As List(Of Node)) As RouteInfo
            'get the number of stops on each path
            Dim route = New StringBuilder()
            Dim totalDistance As Integer = 0
            Dim counter As Integer = 0
            Dim initialNode = New Node()

            For Each stopOver As Node In paths
                ' set route map
                route.AppendFormat("{0} ", stopOver.Name)

                'get distance between each other
                If counter < paths.Count Then
                    If counter > 0 Then
                        'check if first item
                        'compare
                        Dim endPoint As Edge = initialNode.Edges.FirstOrDefault(Function(e) e.NodeDestination.Name = stopOver.Name)

                        If endPoint Is Nothing Then
                            Return Nothing
                        End If

                        'get distance
                        totalDistance += endPoint.Distance
                    End If

                    'replace initial node; will be used to compare with the next node
                    initialNode = stopOver
                    counter += 1
                End If
            Next

            Dim result = New RouteInfo() With { _
                .Path = route.ToString(), _
                .NumberOfStops = paths.Count, _
                .TotalDistance = totalDistance _
            }

            Return result
        End Function
    End Class
End Namespace
